SwiftPackage/Sources/AssessmentModel/Assessment/NodeSerializer.swift

```diff
@@ -0,0 +1,357 @@
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+import Foundation
+import JsonModel
+
+/// `SerializableNode` is the base implementation for `Node` that is serialized using
+/// the `Codable` protocol and the polymorphic serialization defined by this framework.
+///
+public protocol SerializableNode : Node, PolymorphicRepresentable {
+    var serializableType: SerializableNodeType { get }
+}
+
+extension SerializableNode {
+    public var typeName: String { serializableType.stringValue }
+}
+
+/// `SerializableNodeType` is an extendable string enum used by the `SerializationFactory` to
+/// create the appropriate result type.
+public struct SerializableNodeType : TypeRepresentable, Codable, Hashable {
+    
+    public let rawValue: String
+    
+    public init(rawValue: String) {
+        self.rawValue = rawValue
+    }
+    
+    public enum StandardTypes : String, CaseIterable {
+        case assessment, section, overview, instruction, choiceQuestion, simpleQuestion, completion
+        
+        public var nodeType: SerializableNodeType {
+            .init(rawValue: self.rawValue)
+        }
+    }
+    
+    /// List of all the standard types.
+    public static func allStandardTypes() -> [SerializableNodeType] {
+        StandardTypes.allCases.map { $0.nodeType }
+    }
+}
+
+extension SerializableNodeType : ExpressibleByStringLiteral {
+    public init(stringLiteral value: String) {
+        self.init(rawValue: value)
+    }
+}
+
+extension SerializableNodeType : DocumentableStringLiteral {
+    public static func examples() -> [String] {
+        return allStandardTypes().map{ $0.rawValue }
+    }
+}
+
+public final class NodeSerializer : IdentifiableInterfaceSerializer, PolymorphicSerializer {
+    public var documentDescription: String? {
+        """
+        The interface for any `Node` that is serialized using the `Codable` protocol and the
+        polymorphic serialization defined by this framework.
+        """.replacingOccurrences(of: "\n", with: " ").replacingOccurrences(of: "  ", with: "\n")
+    }
+    
+    public var jsonSchema: URL {
+        URL(string: "\(self.interfaceName).json", relativeTo: kSageJsonSchemaBaseURL)!
+    }
+    
+    override init() {
+        self.examples = [
+            AssessmentObject.examples().first!,
+            ChoiceQuestionStepObject.examples().first!,
+            CompletionStepObject.examples().first!,
+            InstructionStepObject.examples().first!,
+            OverviewStepObject.examples().first!,
+            SectionObject.examples().first!,
+            SimpleQuestionStepObject.examples().first!,
+        ]
+    }
+    
+    public private(set) var examples: [Node]
+    
+    public override class func typeDocumentProperty() -> DocumentProperty {
+        .init(propertyType: .reference(SerializableNodeType.documentableType()))
+    }
+    
+    /// Insert the given example into the example array, replacing any existing example with the
```

Comment:
Basically, `example` is a `Set` where there is a one-to-one mapping of "type" keyword to a single class or struct, *but* in order to be in a set, the object must be `Hashable` which is a generic protocol with an associated type. Because of this, I can't use a `Set` and the protocol that actually implements all the goodness has to use an array where each element is unique based on its "type" keyword. So it doesn't matter what the index into the array is, just that there can be only one.

Ideally, we wouldn't do this. Instead, if you added an example with the same serialization "type" keyword as an existing member of the serializer, it would fail *but* Northwestern repurposes our type names for each of their assessments and then has a factor specific to the mapped assessment. So for them "task", "instruction", etc. can mean several different objects. Additionally, these serializers were developed in large part as a path of least code change when migrating to using the Kotlin serialization architecture and naming conventions.